File: src/tree/Solution199.java
// 199. 二叉树的右视图
package tree;

import java.util.*;

public class Solution199 {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        if (root == null) return ans;
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int n = queue.size();
            TreeNode tmp = new TreeNode();
            for (int i = 0; i < n; i++) {
                tmp = queue.poll();
                if (tmp.left != null) queue.offer(tmp.left);
                if (tmp.right != null) queue.offer(tmp.right);
            }
            // tmp is the rightmost node
            ans.add(tmp.val);
        }

        return ans;
    }
}

File: python/trees/Solution199.py
# 199. 二叉树的右视图
from collections import deque
from typing import Optional, List

from trees import TreeNode


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        queue = deque([root])
        res = []
        while queue:
            length = len(queue)
            for _ in range(length):
                tmp = queue.popleft()
                if tmp.left:
                    queue.append(tmp.left)
                if tmp.right:
                    queue.append(tmp.right)
            res.append(tmp.val)
        return res
